lib/checks/aria/aria-allowed-attr-evaluate.js

```diff
@@ -27,21 +28,56 @@ import { getRole, allowedAttr, validateAttr } from '../../commons/aria';
  */
 function ariaAllowedAttrEvaluate(node, options, virtualNode) {
   const invalid = [];
-
   const role = getRole(virtualNode);
   const attrs = virtualNode.attrNames;
   let allowed = allowedAttr(role);
-
   // @deprecated: allowed attr options to pass more attrs.
   // configure the standards spec instead
   if (Array.isArray(options[role])) {
     allowed = uniqueArray(options[role].concat(allowed));
   }
 
+  // TODO: look into memoizing getRole and removing this local cache
+  let tableMap = cache.get('aria-allowed-attr-table');
+  if (!tableMap) {
+    tableMap = new WeakMap();
+    cache.set('aria-allowed-attr-table', tableMap);
+  }
+
+  function validateRowAttrs() {
+    // check if the parent exists otherwise a TypeError will occur (virtual-nodes specifically)
+    if (virtualNode.parent && role === 'row') {
+      const table = closest(
+        virtualNode,
+        'table, [role="treegrid"], [role="table"], [role="grid"]'
+      );
+
+      let tableRole = tableMap.get(table);
+      if (table && !tableRole) {
+        tableRole = getRole(table);
+        tableMap.set(table, tableRole);
+      }
+      if (['table', 'grid'].includes(tableRole) && role === 'row') {
+        return true;
+      }
+    }
+  }
+  // Allows options to be mapped to object e.g. {'aria-level' : validateRowAttrs}
+  const preChecks = Object.assign(
+    {},
+    ...options.ariaAttr.map(key => ({ [key]: validateRowAttrs }))
+  );
+
   if (allowed) {
     for (let i = 0; i < attrs.length; i++) {
       const attrName = attrs[i];
-      if (validateAttr(attrName) && !allowed.includes(attrName)) {
+      if (
+        validateAttr(attrName) &&
+        attrName in preChecks &&
+        preChecks[attrName]()
```

Comment:
Can probably just do this:

```suggestion
        preChecks[attrName]?.()
```